pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	return nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		err := filepath.Walk(repo.GetPath(), func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				return nil
+			}
+			appInfo, skip := r.readApplicationInfo(repoID, repo.GetPath(), filepath.Base(path), registeredAppPaths, false)
+			if skip {
+				return nil
+			}
+			apps = append(apps, appInfo)
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to inspect files under %s: %w", repo.GetPath(), err)
+		}
+	}
+	if len(apps) == 0 {
+		return nil
+	}
+
+	_, err := r.apiClient.ReportUnregisteredApplicationConfigurations(
+		ctx,
+		&pipedservice.ReportUnregisteredApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to put the latest unregistered application configurations: %w", err)
+	}
+	return nil
+}
+
+// updateRegisteredApps sends application configurations that have changed since the last time to the control-plane.
+func (r *Reporter) updateRegisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) (err error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		var headCommit git.Commit
+		headCommit, err = repo.GetLatestCommit(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get the latest commit of %s: %w", repoID, err)
+		}
+		lastScannedCommit, ok := r.lastScannedCommits[repoID]
+		if ok && headCommit.Hash == lastScannedCommit {
+			continue
+		}
+
+		var files []string
+		files, err = repo.ChangedFiles(ctx, lastScannedCommit, headCommit.Hash)
+		if err != nil {
+			return fmt.Errorf("failed to get files those were touched between two commits: %w", err)
+		}
+		if len(files) == 0 {
+			// The case where all changes have been fully reverted.
+			continue
+		}
+		for _, filename := range files {
+			appInfo, skip := r.readApplicationInfo(repoID, repo.GetPath(), filename, registeredAppPaths, true)
+			if skip {
+				continue
+			}
+			apps = append(apps, appInfo)
+		}
+
+		id := repoID
+		defer func() {
+			if err == nil {
+				r.lastScannedCommits[id] = headCommit.Hash
+			}
+		}()
+	}
+	if len(apps) == 0 {
+		return
+	}
+
+	_, err = r.apiClient.UpdateApplicationConfigurations(
+		ctx,
+		&pipedservice.UpdateApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to update application configurations: %w", err)
+	}
+	return
+}
+
+func (r *Reporter) readApplicationInfo(repoID, path, cfgFilename string, registeredAppPaths map[string]struct{}, wantRegistered bool) (appInfo *model.ApplicationInfo, skip bool) {
```

Comment:
To be simple, `registeredAppPaths map[string]struct{}, wantRegistered bool` should not be a part of this function.